app/models/user.rb

```diff
@@ -162,8 +164,50 @@ def remember_me?
     remember_token_expires_at && remember_token_expires_at > Time.zone.now
   end
 
+  def mfa_enabled?
+    !no_mfa?
+  end
+
+  def disable_mfa!
+    no_mfa!
+    self.mfa_seed = ''
+    self.mfa_recovery_codes = []
+    self.last_otp_at = nil
+    save!(validate: false)
+  end
+
+  def enable_mfa!(seed, level)
+    self.mfa_level = level
+    self.mfa_seed = seed
+    self.mfa_recovery_codes = Array.new(10).map { SecureRandom.hex(6) }
+    save!(validate: false)
+  end
+
+  def otp_verified?(otp)
+    if no_mfa? || verify_digit_otp(otp)
+      true
+    else
+      is_recovery = mfa_recovery_codes.include?(otp)
+      if is_recovery
+        mfa_recovery_codes.delete(otp)
+        save!(validate: false)
+      end
+      is_recovery
+    end
+  end
+
   private
 
+  def verify_digit_otp(otp)
+    totp = ROTP::TOTP.new(mfa_seed)
+    last_success = totp.verify_with_drift_and_prior(otp, 30, last_otp_at)
+    if last_success
+      self.last_otp_at = Time.at(last_success).utc.to_datetime
+      save!(validate: false)
+    end
+    !last_success.nil?
```

Comment:
if block here could be changed `return false unless last_success`. IMHO, [guard statements](https://www.thechrisoshow.com/2009/02/16/using-guard-clauses-in-your-ruby-code/) are more readable.